Fix interpreter_exec.jl test

Keno opened the pull request:
This test was supposed to check that we correctly handled PhiNodes in uninferred code in both the interpreter and the compiler. However, the compiler path wasn't actually exercised, because the `inferred=true` part of this forced it to be skipped. Drop that and fix the exposed issues in the compiler where we didn't handle PhiNodes properly.

Keno committed:
Fix interpreter_exec.jl test

This test was supposed to check that we correctly handled PhiNodes in uninferred
code in both the interpreter and the compiler. However, the compiler path wasn't
actually exercised, because the `inferred=true` part of this forced it to be skipped.
Drop that and fix the exposed issues in the compiler where we didn't handle PhiNodes
properly.

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -3304,6 +3304,10 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
             end
             if rt === Bottom
                 ssavaluetypes[currpc] = Bottom
+                # Special case: Union typed PhiNodes do not error (but must also be unused)
```
```suggestion
                # Special case: Bottom-typed PhiNodes do not error (but must also be unused)
```

vtjnash replied:
Does this apply to PhiC also?

aviatesk commented on test/compiler/interpreter_exec.jl:
```diff
@@ -20,10 +20,9 @@ let m = Meta.@lower 1 + 1
-    src.ssavaluetypes = Any[ Any for _ = 1:nstmts ]
+    src.ssavaluetypes = nstmts
     src.ssaflags = fill(UInt8(0x00), nstmts)
     src.codelocs = fill(Int32(1), nstmts)
```
```suggestion
    src.codelocs = fill(Int32(1), nstmts)
    @assert !src.inferred
```
instead?

aviatesk committed:
handle unassigned `PhiNode` values in `find_ssavalue_uses`

aviatesk commented on base/compiler/utilities.jl:
```diff
@@ -433,13 +433,16 @@ function find_ssavalue_uses(e::Expr, uses::Vector{BitSet}, line::Int)
 
-function find_ssavalue_uses(e::PhiNode, uses::Vector{BitSet}, line::Int)
-    for val in e.values
+function find_ssavalue_uses!(e::PhiNode, uses::Vector{BitSet}, line::Int)
+    values = e.values
+    for i = 1:length(values)
+        isassigned(values) || continue
+        val = values[i]
```
Implemented a fix for the segmentation fault occurring in `test/compiler/ssair.jl`.

vtjnash commented on base/compiler/utilities.jl:
```diff
@@ -433,13 +433,16 @@ function find_ssavalue_uses(e::Expr, uses::Vector{BitSet}, line::Int)
+function find_ssavalue_uses!(e::PhiNode, uses::Vector{BitSet}, line::Int)
+    values = e.values
+    for i = 1:length(values)
+        isassigned(values) || continue
```
```suggestion
        isassigned(values, i) || continue
```
?

vtjnash committed:
Apply suggestions from code review

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>

vtjnash committed:
Update base/compiler/abstractinterpretation.jl

aviatesk committed:
refactors `interpreter_exec.jl` further (#53246)

This commit refines `interpreter_exec.jl` further, on top of #53218. In
detail, by using `Base.Experimental.@compiler_options compile=min`, it's
now possible to mimic the effect of `julia --compile=[min|yes]`. This
commit is leveraging it and includes `interpreter_exec.jl` from
`test/compiler/ssair.jl`, allowing us to get better stack traces and
test summaries.

In addition, I've introduced several tests concerning the `src.inferred`
values and have relocated the tests from #47066 within
`interpreter_exec.jl` to `test/compiler/irpasses.jl`, since it is
related to Julia-level compilation and not to interpreter execution.

aviatesk committed:
better argument order
